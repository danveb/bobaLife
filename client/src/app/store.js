import { combineReducers, configureStore } from "@reduxjs/toolkit"; 
import authReducer from "../redux/auth/authSlice"; 
import productsReducer from "../redux/products/productsSlice"; 
import cartReducer from "../redux/cart/cartSlice"; 
import {
    persistStore,
    persistReducer,
    FLUSH,
    REHYDRATE,
    PAUSE,
    PERSIST,
    PURGE,
    REGISTER,
} from 'redux-persist';
import storage from "redux-persist/lib/storage"; 

// Perhaps combine all reducers and persist? 
// even persist user so we don't save user data in localStorage?
// combineReducers from rootReducer
const rootReducer = combineReducers({
    auth: authReducer, 
    products: productsReducer, 
    cart: cartReducer, 
}); 

const persistConfig = {
    key: 'root',
    storage,
}
  
// const persistedReducer = persistReducer(persistConfig, cartReducer)
const persistedReducer = persistReducer(persistConfig, rootReducer)

export const store = configureStore({
    // reducer: {
    //     auth: authReducer, 
    //     products: productsReducer, 
    //     cart: persistedReducer, 
    // },
    reducer: persistedReducer, 
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: {
                ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
            },
        }),
}); 

export const persistor = persistStore(store) 