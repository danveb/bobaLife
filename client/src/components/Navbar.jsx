import { Link, useNavigate } from "react-router-dom"; 
import { useSelector, useDispatch } from "react-redux"; 
import { logout, reset } from "../redux/auth/authSlice"; 
import { linkItems } from "../constants/data";
import { GiBoba } from "react-icons/gi"; 
import "./Navbar.scss"; 

const Navbar = () => {
    // useNavigate
    const navigate = useNavigate(); 

    // useDispatch 
    const dispatch = useDispatch(); 

    // useSelector
    const { user } = useSelector((state) => state.auth); 

    const handleLogout = () => {
        dispatch(logout()); 
        dispatch(reset());
        navigate("/"); 
    }

    return (
        <div className="nav-container">
            <div className="nav-wrapper">
                <div className="nav-left">
                    <a href="#home">
                        <GiBoba />
                    </a>
                </div>
                <div className="nav-center">
                    <a href="#home">
                        <h1>bobaLife</h1>
                    </a>
                </div>
                <div className="nav-right">
                    <ul>
                    {linkItems.map((item) => (
                        <li key={item.id} className="nav-menuItem">
                            <a href={item.url}>
                                {item.text}
                            </a>
                        </li>
                    ))}
                    { user ? (
                        <button className="btn-link" onClick={handleLogout}>
                            Logout
                        </button>
                    ) : (
                        <>
                        <li className="nav-menuItem">
                            <Link to="/login">Login</Link>
                        </li>
                        <li className="nav-menuItem">
                            <Link to="/register">Register</Link>
                        </li>
                        </>
                    )}
                    </ul>
                </div>
            </div>
        </div>
    )
}

export default Navbar