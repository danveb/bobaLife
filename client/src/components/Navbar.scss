.nav-container {
    width: 100%; 
    height: 60px; 
    position: sticky; 
    top: 0; 
    z-index: 5; 
    background-color: #fff; 
    padding: 10px; 
}

.nav-wrapper {
    display: flex; 
    justify-content: space-between; 
    align-items: center; 
}

.nav-left {
    flex: 1; 
    display: flex; 
    justify-content: flex-start; 
    align-items: center; 

    svg {
        width: 40px; 
        height: 40px; 
        border-radius: 50%; 
        color: #fff; 
        background-color: #55acee; 
    }
}

.nav-center {
    flex: 1; 
    display: flex; 
    justify-content: center; 
    align-items: center; 

    a {
        text-decoration: none; 
        color: black; 

        h1 {
            text-align: center; 
        }
    }

    @media screen and (max-width: 768px) {
        display: none; 
    }
}

.nav-right {
    flex: 1; 
    display: flex; 
    justify-content: flex-end; 
    align-items: center; 

    ul {
        display: flex; 
        justify-content: center; 
        align-items: center; 
    }

    @media screen and (max-width: 768px) {
        display: none; 
    }
}

.nav-menuItem {
    list-style: none; 
    cursor: pointer; 
    margin-left: 20px; 

    a {
        text-decoration: none; 
        color: black; 
    }
}

.btn-link {
    border: none; 
    margin-left: 20px; 
    background: #fff;
    color: #000;
    font-size: 16px;
    cursor: pointer;
}

// NAV MOBILE
.nav-mobile {
    width: 35px; 
    height: 35px; 
    border-radius: 50%; 
    // position: relative; 
    display: flex; 
    justify-content: flex-between; 
    align-items: center; 

    svg {
        width: 90%;  
        height: 90%; 
    }

    div {
        position: fixed; 
        top: 0; 
        bottom: 0; 
        right: 0; 
        z-index: 10; 
        padding: 10px; 
        width: 50%; 
        height: 100vh; 
        display: flex; 
        justify-content: flex-end; 
        align-items: flex-end; 
        flex-direction: column; 
        // background-color: rgb(35, 109, 90); 
        background-color: #fff; 
        box-shadow: 0px 0px 20px rgba(168, 168, 168, 0.15);

        svg {
            width: 35px; 
            height: 35px; 
            margin: 0.5rem 0.7rem; 
        }

        ul {
            list-style: none; 
            margin: 0; 
            padding: 0; 
            height: 100%; 
            width: 100%; 
            display: flex; 
            justify-content: flex-start; 
            align-items: flex-start; 
            flex-direction: column; 

            li {
                margin: 1rem; 

                a {
                    color: #000; 
                    text-decoration: none; 
                    flex-direction: column;  
                    transition: all 0.3s ease-in-out; 
                }
            }

            button {
                margin: 1rem; 
                background-color: #fff;
                border: none; 
                color: #000;
                font-size: 16px;
                cursor: pointer;
            }
        }

        @media screen and (min-width: 768px) {
            display: none; 
        }
    }

    @media screen and (min-width: 768px) {
        display: none; 
    }
}