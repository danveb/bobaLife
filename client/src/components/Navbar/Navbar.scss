.navbar {
    width: 100%; 
    position: relative; 
    top: 0;
    z-index: 10; 
    padding: 24px 24px 0 24px; 
    
    .nav-wrapper {
        width: 100%; 
        height: 55px; 
        display: flex; 
        flex-direction: row; 
        background-color: var(--languid-lavender); 
        padding: 15px 30px; 
        transition: all 1s ease; 
    
        .nav-left {
            flex: 1; 
            display: flex; 
            justify-content: flex-start; 
            align-items: center; 

            .logo {
                text-decoration: none; 
                color: var(--raisin-black); 
                font-size: 24px; 
            }
        }
    
        .nav-center {
            flex: 1; 
            display: flex; 
            justify-content: center; 
            align-items: center; 

            .nav-links {
                display: flex; 
                justify-content: center; 
                align-items: center; 
                
                li {
                    list-style: none; 
                    margin: 0 15px; 
                    cursor: pointer; 

                    a {
                        text-decoration: none; 
                        color: var(--raisin-black); 
                        font-size: 16px; 
                    }
                }

                @media screen and(max-width: 768px) {
                    display: none; 
                }
            }
        }
    
        .nav-right {
            flex: 1; 
            display: flex; 
            justify-content: flex-end; 
            align-items: center; 
            gap: 15px; 

            svg {
                display: flex; 
                justify-content: center; 
                align-items: center; 
                height: 20px; 
                width: 20px; 
                color: var(--raisin-black); 
                cursor: pointer; 
            }

            .hamburger {
                display: flex; 
                flex-direction: column; 
                justify-content: space-between; 
                width: 18px; // 18px // 24px
                height: 15px; // 15px // 20px
                cursor: pointer; 
    
                span {
                    width: 100%; 
                    height: 2px; 
                    background-color: var(--raisin-black); 
                    transform-origin: left; 
                    transition: all 0.7s ease; 
                }
    
                @media screen and(min-width: 768.9px) {
                    display: none; 
                }
            }
        }

        &.active {
            background-color: var(--cadet-blue-crayola); 
            color: var(--alice-blue); 
    
            .hamburger {
    
                span {
                    &:first-child {
                        background-color: var(--raisin-black); 
                        transform: rotate(45deg); 
                    }
    
                    &:nth-child(2) {
                        opacity: 0; // hide color
                    }
    
                    &:last-child {
                        background-color: var(--raisin-black); 
                        transform: rotate(-45deg); 
                    }
                }
            }
        }
    }
}