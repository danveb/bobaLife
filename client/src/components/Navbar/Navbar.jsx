import { Link, useNavigate } from "react-router-dom"; 
import { useSelector, useDispatch } from "react-redux"; 
import { logout, reset } from "../../redux/auth/authSlice"; 
import { clearCart } from "../../redux/cart/cartSlice";
import { links } from "../../constants/links";
import { Badge } from "@mui/material"; 
import { AccountCircle, ShoppingBasket, PowerSettingsNew } from "@mui/icons-material"
import "./Navbar.scss"; 
import { toast } from "react-toastify"; 

const Navbar = ({ menuOpen, setMenuOpen }) => {
    // useNavigate
    const navigate = useNavigate(); 

    // useDispatch 
    const dispatch = useDispatch(); 

    // useSelector
    const { user } = useSelector((state) => state.auth); 
    const { cartItems } = useSelector((state) => state.cart); 

    const handleLogout = () => {
        dispatch(logout()); 
        dispatch(reset());
        dispatch(clearCart()); 
        toast.info(`Sad to see you go ${user.username} ðŸ˜­ `, {
            position: "top-center", 
            autoClose: 2000, 
            pauseOnHover: false, 
        });
        navigate("/"); 
    }; 

    const cartQty = () => {
        return cartItems.reduce((acc, item) => acc + item.qty, 0); 
    }; 

    return (
        <div className="navbar">
            <div className={"nav-wrapper " + (menuOpen && "active")}>
                <div className="nav-left">
                    <Link to="/" className="logo" onClick={() => setMenuOpen(false)}>bobaLife</Link>
                </div>
                <div className="nav-center">
                    <ul className="nav-links">
                        {links.map((link) => (
                            <li key={link.id}>
                                <Link to={`${link.path}`}>
                                    {link.text}
                                </Link>
                            </li>
                        ))}
                    </ul>
                </div>
                <div className="nav-right">
                    {user ? (
                        <PowerSettingsNew onClick={handleLogout} />
                        ) : (
                            <Link to="/login"><AccountCircle onClick={() => setMenuOpen(false)}/></Link>
                        )}
                    <Link to="/cart">
                        <Badge color="primary" badgeContent={cartQty()}>
                            <ShoppingBasket onClick={() => setMenuOpen(false)}/>
                        </Badge>
                    </Link>
                    <div className="hamburger" onClick={() => setMenuOpen(!menuOpen)}>
                        <span></span>
                        <span></span>
                        <span></span>
                    </div>
                </div>
            </div>
        </div>
    )
}

export default Navbar