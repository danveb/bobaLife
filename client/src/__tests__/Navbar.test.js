import { render, screen } from "@testing-library/react"; 
import { store } from "../app/store"; 
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import Navbar from "../components/Navbar/Navbar"; 

describe("Navbar component", () => {
    // smoke test
    test("renders without crashing", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Navbar />
                </BrowserRouter>
            </Provider>
        ); 
    }); 

    // snapshot test
    test("matches snapshot", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Navbar />
                </BrowserRouter>
            </Provider>
        ); 
        expect(asFragment()).toMatchSnapshot(); 
    });

    // getByText
    test("should display app logo", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Navbar />
                </BrowserRouter>
            </Provider>
        ); 
        const logo = screen.getByText("bobaLife", { exact: true })
        expect(logo).toBeInTheDocument(); 
    });
});