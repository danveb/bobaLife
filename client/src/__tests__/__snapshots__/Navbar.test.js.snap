// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="navbar"
  >
    <div
      class="nav-wrapper undefined"
    >
      <div
        class="nav-left"
      >
        <a
          class="logo"
          href="/"
        >
          bobaLife
        </a>
      </div>
      <div
        class="nav-center"
      >
        <ul
          class="nav-links"
        >
          <li>
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              href="/about"
            >
              About
            </a>
          </li>
          <li>
            <a
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li>
            <a
              href="/shop"
            >
              Shop
            </a>
          </li>
        </ul>
      </div>
      <div
        class="nav-right"
      >
        <a
          href="/login"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AccountCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20z"
            />
          </svg>
        </a>
        <a
          href="/cart"
        >
          <span
            class="MuiBadge-root BaseBadge-root css-1c32n2y-MuiBadge-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ShoppingBasketIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m17.21 9-4.38-6.56c-.19-.28-.51-.42-.83-.42-.32 0-.64.14-.83.43L6.79 9H2c-.55 0-1 .45-1 1 0 .09.01.18.04.27l2.54 9.27c.23.84 1 1.46 1.92 1.46h13c.92 0 1.69-.62 1.93-1.46l2.54-9.27L23 10c0-.55-.45-1-1-1h-4.79zM9 9l3-4.4L15 9H9zm3 8c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2z"
              />
            </svg>
            <span
              class="MuiBadge-badge MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary BaseBadge-badge BaseBadge-invisible css-1pi4uwz-MuiBadge-badge"
            />
          </span>
        </a>
        <div
          class="hamburger"
        >
          <span />
          <span />
          <span />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
